src/NitroSharp/Content/ArchiveManager.cs

```diff
@@ -0,0 +1,255 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.MemoryMappedFiles;
+using System.Text;
+
+namespace NitroSharp.Content
+{
+    internal interface IArchiveFile : IDisposable
+    {
+        public void OpenArchive();
+        public Stream OpenStream(string path);
+        public bool Contains(string path);
+    }
+
+    internal class VFSNode : IDisposable
+    {
+        public List<IArchiveFile>? MountedArchives { get; set; }
+        public Dictionary<string, VFSNode>? Children { get; set; }
+
+        private bool _disposed = false;
+
+        ~VFSNode()
+        {
+            Dispose(false);
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public virtual void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (disposing)
+                {
+                    if (Children != null)
+                    {
+                        foreach (KeyValuePair<string, VFSNode> pair in Children)
+                        {
+                            pair.Value.Dispose();
+                        }
+                    }
+
+                    if (MountedArchives != null)
+                    {
+                        foreach (IArchiveFile mountedArchive in MountedArchives)
+                        {
+                            mountedArchive.Dispose();
+                        }
+                    }
+                }
+                _disposed = true;
+            }
+        }
+    }
+
+    public class ArchiveException : Exception
+    {
+        public ArchiveException()
+            : base("Unable to open the archive")
+        {
+        }
+
+        public ArchiveException(string format, string message)
+            : base($"{format} : {message}")
+        {
+        }
+
+    }
+
+    internal class ArchiveManager : IDisposable
+    {
+        public static readonly string DefaultEncoding = "shift-jis";
+
+        private VFSNode _root;
+        private Encoding _encoding;
+
+        private bool _disposed = false;
+
+        public ArchiveManager(Encoding? encoding = null)
+        {
+            if (encoding == null)
+            {
+                try
+                {
+                    encoding = Encoding.GetEncoding(DefaultEncoding);
+                }
+                catch (ArgumentException)
+                {
+                    Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
```

Comment:
Consider calling `Encoding.RegisterProvider` in the static constructor instead of catching `ArgumentException`. See [SourceText.cs](https://github.com/CommitteeOfZero/nitrosharp/blob/b5229145a0129624103c6d19fd639623be73c056/src/NitroSharp.NsScript/SourceText.cs#L13).

Reply:
I did it that way because it might have already been called by the bytecode compiler if it ran, but you're right, running it twice in that situation isn't a problem and makes the code clearer.